server/console_api_explorer.go

```diff
@@ -0,0 +1,268 @@
+		if cval != nil {
+			m := new(jsonpb.Marshaler)
+			j, err = m.MarshalToString(cval.(proto.Message))
+			if err != nil {
+				s.logger.Error("Error serializing method response body.", zap.String("method", in.Method), zap.Error(err))
+				return nil, status.Error(codes.Internal, "Error serializing method response body.")
+			}
+		}
+		return &console.CallApiEndpointResponse{
+			Body: j,
+		}, nil
+	}
+
+}
+
+func (s *ConsoleServer) extractApiCallContext(ctx context.Context, in *console.CallApiEndpointRequest, userIdOptional bool) (context.Context, error) {
+	var callCtx context.Context
+	if strings.HasPrefix(in.Method, "Authenticate") {
+		callCtx = ctx
+	} else if in.UserId == "" {
+		if !userIdOptional {
+			s.logger.Error("Error calling a built-in RPC function without a user_id.", zap.String("method", in.Method))
+			return nil, status.Error(codes.InvalidArgument, "Built-in RPC functions require a user_id.")
+		} else {
+			callCtx = ctx
+		}
+	} else {
+		row := s.db.QueryRowContext(ctx, "SELECT username FROM users WHERE id = $1", in.UserId)
+		var dbUsername string
+		userUUID, err := uuid.FromString(in.UserId)
+		if err != nil {
+			s.logger.Error("Invalid user uuid.", zap.String("method", in.Method))
+			return nil, status.Error(codes.InvalidArgument, "Invalid user uuid.")
+		}
+		err = row.Scan(&dbUsername)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				s.logger.Error("User id not found.", zap.String("method", in.Method))
+				return nil, status.Error(codes.InvalidArgument, "User id not found.")
+			}
+			s.logger.Error("Error looking up user account.", zap.String("method", in.Method), zap.Error(err))
+			return nil, status.Error(codes.Internal, "Error looking up user account.")
+		}
+		callCtx = context.WithValue(context.WithValue(ctx, ctxUserIDKey{}, userUUID), ctxUsernameKey{}, dbUsername)
+	}
+	return callCtx, nil
+
+}
+
+func (s *ConsoleServer) ListApiEndpoints(ctx context.Context, _ *empty.Empty) (*console.ApiEndpointList, error) {
+	if s.methodListCache == nil {
+		s.methodListCache = make([]*console.ApiEndpointDescriptor, 0)
+		apiType := reflect.TypeOf(s.api)
+		for i := 0; i < apiType.NumMethod(); i++ {
+			method := apiType.Method(i)
+			if method.Type.NumIn() != 3 || method.Type.NumOut() != 2 {
+				continue
+			}
+			if method.Name == "Healthcheck" {
+				continue
+			}
+			if method.Name == "RpcFunc" {
+				continue
+			}
+			request := method.Type.In(2)
+			if request.Kind() == reflect.Ptr {
+				request = request.Elem()
+			}
+			endpoint := &console.ApiEndpointDescriptor{
+				Method:       method.Name,
+				BodyTemplate: reflectProtoMessageAsJsonTemplate(request),
+			}
+
+			s.methodListCache = append(s.methodListCache, endpoint)
+		}
+	}
+
+	rpcs := make([]string, 0)
+	for _, rpc := range s.runtimeInfo.LuaRpcFunctions {
+		rpcs = append(rpcs, rpc)
+	}
+	for _, rpc := range s.runtimeInfo.GoRpcFunctions {
+		rpcs = append(rpcs, rpc)
+	}
+	for _, rpc := range s.runtimeInfo.JavaScriptRpcFunctions {
```

Comment:
I can see a small problem case here: if there are multiple functions registered for the same RPC ID, in different runtimes, they can't each be invoked individually. The server picks one to prioritise - the Go runtime usually.

So returning all of them separately here might be misleading. Perhaps we should return as part of the `runtimeInfo` the "merged" list the server already has for this?

Reply:
As an added bonus in the console runtime view we should mark which functions are overridden by a function registered with the same name but in a "higher priority" runtime.

Reply:
So this is something that will need some thinking and addition to the protos, i'll come up with something. If calling is done with a prioritization than we should look at the prirority conditions and exactly replicated them in the explorer to display the one that is going to be excuted.